util_function_groups/column_operations.R

```diff
@@ -3,6 +3,219 @@
 source("~/Sites/gymnast/util.R", chdir = FALSE)
 source("~/Sites/gymnast/util_qualtrics_cleaning.R", chdir = FALSE)
 
+
+############################### rename_columns function  #################################################
+
+rename_columns <- function(data, mapping) {
+  # This function takes a data set and a mapping (which can be a list or data frame)
+  # with columns "old_name" and "new_name". It uses util.recode to turn the old names
+  # into the new names, making sure not to break any R name rules.
+  
+  # Patterns for forbidden names
+  whitespace_pattern <- "\\s"
+  starting_w_digit_pattern <- "^\\.?[0-9]"
+  naughty_words <- c("if", "else", "repeat", "while", "function", "for",
+                     "in", "next", "break", "TRUE", "FALSE", "NULL", "Inf",
+                     "NaN", "NA", "NA_integer_", "NA_real_", "NA_complex_", "NA_character_", NA)
+  
+  # if the new names in the mapping match any evil patterns, then complain
+  if( any(mapping$new_name %in% naughty_words) |
+      any(grepl(whitespace_pattern, mapping$new_name)) |
+      any(grepl(starting_w_digit_pattern, mapping$new_name))) {
+    stop("Invalid new names are present among: ",paste(mapping$new_name, collapse =" "))
+  }
+  
+  # if old and new names are not the same length, then complain
+  if(length(mapping$old_name) != length(mapping$new_name)) {
+    stop("Error! Old names and new names are not the same length in your mapping for renaming.")
+  }
+  
+  names(data) <- util.recode(names(data),
+                             mapping$old_name,
+                             mapping$new_name)
+  
+  # if new column names in the final data have duplicates, then complain
+  if (any(duplicated(names(data)))) {
+    stop("Error! Duplicated names in your renamed data. Review your mapping for renaming.")
+  }
+  
+  return(data)
+}
+
+################################ testing for rename_columns  ################################################
+
+######################## TEST 1: handling forbidden names
+
+
+test_forbidden_names <- function() {
+  
+  my_mapping <- list(
+    old_name = c("foo", "bar", "baz","blerp"),
+    new_name = c("newfoo", "newbar", "newbaz", NA)
+  )
+  
+  d <- data.frame(
+    foo = c(1,2,3,4,5),
+    bar = c(6,7,8,9,10),
+    baz = c(11,12,13,14,15),
+    blerp = c(16,17,18,19,20)
+  )
+  
+  success <- FALSE
+  
+  tryCatch(
+    expr = {
+      rename_columns(d,my_mapping)
+    },
+    error = function(e) {
+      success <<- TRUE
+    }
+  )
+  return(success)
```

Comment:
Let's add a todo in the comments about abstracting this "assert that the function raises an error" code.